src/hotspot/share/utilities/resizeableResourceHash.hpp

```diff
@@ -0,0 +1,133 @@
+#define SHARE_UTILITIES_RESIZEABLERESOURCEHASH_HPP
+
+#include "utilities/resourceHash.hpp"
+
+template<
+    typename K, typename V,
+    ResourceObj::allocation_type ALLOC_TYPE,
+    MEMFLAGS MEM_TYPE>
+class ResizeableResourceHashtableStorage : public ResourceObj {
+  using Node = ResourceHashtableNode<K, V>;
+
+protected:
+  unsigned _table_size;
+  Node** _table;
+
+  ResizeableResourceHashtableStorage(unsigned table_size) {
+    _table_size = table_size;
+    _table = alloc_table(table_size);
+  }
+
+  ~ResizeableResourceHashtableStorage() {
+    if (ALLOC_TYPE == C_HEAP) {
+      FREE_C_HEAP_ARRAY(Node*, _table);
+    }
+  }
+
+  Node** alloc_table(unsigned table_size) {
+    Node** table;
+    if (ALLOC_TYPE == C_HEAP) {
+      table = NEW_C_HEAP_ARRAY(Node*, table_size, MEM_TYPE);
+    } else {
+      table = NEW_RESOURCE_ARRAY(Node*, table_size);
+    }
+    memset(table, 0, table_size * sizeof(Node*));
+    return table;
+  }
+
+  unsigned table_size() const {
+    return _table_size;
+  }
+
+  Node** table() const {
+    return _table;
+  }
+};
+
+template<
+    typename K, typename V,
+    ResourceObj::allocation_type ALLOC_TYPE = ResourceObj::RESOURCE_AREA,
+    MEMFLAGS MEM_TYPE = mtInternal,
+    unsigned (*HASH)  (K const&)           = primitive_hash<K>,
+    bool     (*EQUALS)(K const&, K const&) = primitive_equals<K>
+    >
+class ResizeableResourceHashtable : public ResourceHashtableBase<
+    ResizeableResourceHashtableStorage<K, V, ALLOC_TYPE, MEM_TYPE>,
+    K, V, HASH, EQUALS, ALLOC_TYPE, MEM_TYPE> {
+  unsigned _max_size;
+
+  using BASE = ResourceHashtableBase<ResizeableResourceHashtableStorage<K, V, ALLOC_TYPE, MEM_TYPE>, K, V, HASH, EQUALS, ALLOC_TYPE, MEM_TYPE>;
+  using Node = ResourceHashtableNode<K, V>;
+
+public:
+  ResizeableResourceHashtable(unsigned size, unsigned max_size = 0)
+  : BASE(size), _max_size(max_size) {
+    assert(size <= 0x3fffffff && max_size <= 0x3fffffff, "avoid overflow in resize");
+  }
+
+  bool maybe_grow(int load_factor = 8) {
+    unsigned old_size = BASE::_table_size;
+    if (old_size >= _max_size) {
+      return false;
+    }
+    if (BASE::number_of_entries() / int(old_size) > load_factor) {
+      unsigned new_size = MIN2<unsigned>(old_size * 2, _max_size);
+      resize(old_size, new_size);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  void resize(unsigned old_size, unsigned new_size) {
+    Node** old_table = BASE::_table;
+    Node** new_table = BASE::alloc_table(new_size);
+
+    Node* const* bucket = old_table;
+    while (bucket < &old_table[old_size]) {
+      Node* node = *bucket;
+      while (node != NULL) {
+        Node* next = node->_next;
+        unsigned hash = HASH(node->_key);
+        unsigned index = hash % new_size;
+
+        node->_next = new_table[index];
+        new_table[index] = node;
+
+        node = next;
+      }
+      ++bucket;
+    }
+
+    FREE_C_HEAP_ARRAY(Node*, old_table);
```

Comment:
If you resource allocated old_table, this doesn't seem right.
It seems like the ResizeableHashtable should be a CHeapHashtable and it's always resizeable.
And the code either has a fixed size ResourceHashtable or a variable/resizeable CHeapHashtable.

Reply:
Fixed. The `FREE_C_HEAP_ARRAY` should be called only if the table is c-heap allocated.

Per our off-line conversation,  let's leave the 4 combination of {res, c-heap} x {fixed, resizeable}, and live with the bad naming for now (even `ResourceHashtable` is misleading since it can be resource allocated). We should try to start getting rid of the old `Hashtable` classes. Eventually, we can clean up the naming with something like:

- ResizeableResourceHashtable -> Hashtable
- ResourceHashtable -> FixedHashtable

Reply:
Ok, thank you for fixing this.  Yes, these are better names.  We can fix all the uses when we agree on some convention and when we replace the utilities/hashtable hashtable and KVHashtable.